iroh-dns-server/src/store/signed_packets.rs

```diff
@@ -14,7 +16,121 @@ const SIGNED_PACKETS_TABLE: TableDefinition<&SignedPacketsKey, &[u8]> =
 
 #[derive(Debug)]
 pub struct SignedPacketStore {
-    db: Arc<Database>,
+    send: mpsc::Sender<Message>,
+    cancel: CancellationToken,
+    _task: AbortOnDropHandle<()>,
+}
+
+impl Drop for SignedPacketStore {
+    fn drop(&mut self) {
+        self.cancel.cancel();
+    }
+}
+
+enum Message {
+    Upsert {
+        packet: SignedPacket,
+        res: oneshot::Sender<bool>,
+    },
+    Get {
+        key: PublicKeyBytes,
+        res: oneshot::Sender<Option<SignedPacket>>,
+    },
+    Remove {
+        key: PublicKeyBytes,
+        res: oneshot::Sender<bool>,
+    },
+}
+
+struct Actor {
+    db: Database,
+    recv: mpsc::Receiver<Message>,
+    cancel: CancellationToken,
+    max_batch_size: usize,
+    max_batch_time: Duration,
+}
+
+impl Actor {
+    async fn run(self) {
+        match self.run0().await {
+            Ok(()) => {}
+            Err(e) => {
+                tracing::error!("packet store actor failed: {:?}", e);
+            }
+        }
+    }
+
+    async fn run0(mut self) -> anyhow::Result<()> {
+        loop {
+            let transaction = self.db.begin_write()?;
+            let mut tables = Tables::new(&transaction)?;
+            let timeout = tokio::time::sleep(self.max_batch_time);
+            tokio::pin!(timeout);
+            loop {
+                for _ in 0..self.max_batch_size {
+                    tokio::select! {
+                        _ = self.cancel.cancelled() => {
+                            drop(tables);
+                            transaction.commit()?;
+                            return Ok(());
+                        }
+                        _ = &mut timeout => break,
```

Comment:
Wouldn't this mean you're dropping the current `transaction`, and thus not committing it?
Shouldn't you instead commit the transaction here, then break?

Reply:
Yeah, forgot that. Also, there is a loop in a loop in a loop. It should be only 2 levels... 😊